ntoskrnl/config/cmkcbncb.c

```diff
@@ -1134,6 +1134,209 @@ DelistKeyBodyFromKCB(IN PCM_KEY_BODY KeyBody,
+ * Builds an array of KCBs and lockes them. Whether these
+ * KCBs are locked exclusively by calling thread or in shared
+ * mode is defined by the KcbLockFlags parameter.
+ *
+ * @param[in] HashCacheStack
+ * A pointer to a hash cache stack. This stack parameter
+ * stores the convkey hashes of interested KCBs of a
+ * key path name that need to be locked. This parameter
+ * must not be NULL!
+ *
+ * @param[in] KcbLockFlags
+ * Define a lock flag to lock the KCBs. Consult the
+ * CmpLockKcbArray documentation for more information.
+ *
+ * @param[in] Kcb
+ * A pointer to a key control block to be given. This
+ * KCB is included in the array for locking, that is,
+ * given by the CmpParseKey from the parser object.
+ * This parameter must not be NULL!
+ *
+ * @param[in,out] OuterStackArray
+ * A pointer to an outer stack array. This array acts like
+ * an auxiliary array to store the KCB elements for this
+ * function. In return the function gives an array of
+ * locked KCBs, this paramter must not be NULL!
+ *
+ * @param[in] TotalRemainingSubkeys
+ * The number of total remaining subkey levels.
+ *
+ * @param[in] MatchRemainSubkeyLevel
+ * The number of remaining subkey levels that match.
+ *
+ * @return
+ * Returns a pointer to an array of KCBs that have been
+ * locked.
+ *
+ * @remarks
+ * The caller HAS THE RESPONSIBILITY to unlock the KCBs
+ * after the necessary operations are done!
+ */
+PULONG
+NTAPI
+CmpBuildAndLockKcbArray(
+    _In_ PCM_HASH_CACHE_STACK HashCacheStack,
+    _In_ ULONG KcbLockFlags,
+    _In_ PCM_KEY_CONTROL_BLOCK Kcb,
+    _Inout_ PULONG OuterStackArray,
+    _In_ ULONG TotalRemainingSubkeys,
+    _In_ ULONG MatchRemainSubkeyLevel)
+{
+    ULONG KcbIndex = 1, HashStackIndex = 0;
+    PULONG LockedKcbs = NULL;
+    PCM_KEY_CONTROL_BLOCK ParentKcb;
+
+    /* These parameters are expected */
+    ASSERT(HashCacheStack != NULL);
+    ASSERT(Kcb != NULL);
+    ASSERT(OuterStackArray != NULL);
+
+    /* Validate the key path */
+    ParentKcb = Kcb->ParentKcb;
+    CmpValidateKeyPathForKcbArray(ParentKcb,
+                                  TotalRemainingSubkeys,
+                                  MatchRemainSubkeyLevel);
+
+    /* Add the parent to KCBs to lock */
+    if ((ParentKcb) && (TotalRemainingSubkeys == MatchRemainSubkeyLevel))
+    {
+        OuterStackArray[KcbIndex++] = ParentKcb->ConvKey;
+    }
+
+    /* Add the current KCB */
+    OuterStackArray[KcbIndex++] = Kcb->ConvKey;
+
+    /* Loop over the hash stack and grab the hashes for locking */
+    while (HashStackIndex < TotalRemainingSubkeys)
+    {
+        OuterStackArray[KcbIndex++] = HashCacheStack[HashStackIndex].ConvKey;
+
+        /* Go to the next hash */
+        HashStackIndex++;
+
+        /*
+         * Make sure we don't blow up the outer stack as we
+         * are adding up hashes for KCBs to lock.
+         */
+        if (KcbIndex >= CMP_HASH_STACK_LIMIT)
+        {
+            DPRINT1("Limit to store KCBs for locking has been REACHED\n");
```

Comment:
I think this should be impossible because keys nested that deeply are forbidden? I.e. this can probably be an ASSERT?